From f954281511aa85eafeaf7242aa82d4a2d00a7950 Mon Sep 17 00:00:00 2001
From: zhangchengceng <zhangchengceng1@huawei.com>
Date: Wed, 1 Feb 2023 09:10:33 +0800
Subject: [PATCH 556/905] [Huawei] drivers: hooks: support vendor hook without
 tracepoints

Offering: HULK
ont inclusion
category: feature
bugzilla: 188338

-------------------------------

Support vendor hook without CONFIG_TRACEPOINTS enabled. Keep
the interfaces unchanged.

Signed-off-by: zhangchengceng <zhangchengceng1@huawei.com>
---
 drivers/hooks/Kconfig              | 12 ++++-
 drivers/hooks/vendor_hooks.c       | 34 ++++++------
 include/trace/hooks/vendor_hooks.h | 83 ++++++++++++++++++++++++++++--
 3 files changed, 108 insertions(+), 21 deletions(-)

diff --git a/drivers/hooks/Kconfig b/drivers/hooks/Kconfig
index 95a783871708..bd55ae4aa5c0 100644
--- a/drivers/hooks/Kconfig
+++ b/drivers/hooks/Kconfig
@@ -3,13 +3,23 @@ menu "Vendor Hooks"
 
 config VENDOR_HOOKS
         bool "Vendor Hooks"
-        depends on TRACEPOINTS
+        default n
+        select VENDOR_HOOKS_LITE if !TRACEPOINTS
         help
           Enable vendor hooks implemented as tracepoints
 
           Allow vendor modules to attach to tracepoint "hooks" defined via
           DECLARE_HOOK or DECLARE_RESTRICTED_HOOK.
 
+config VENDOR_HOOKS_LITE
+        bool "Vendor Hooks without tracepoints"
+        depends on !TRACEPOINTS
+        help
+          Enable vendor hooks implemented without tracepoints
+
+          Allow vendor modules to attach to lite "hooks" defined via
+          DECLARE_HOOK or DECLARE_RESTRICTED_HOOK.
+
 config VENDOR_BOND_HOOKS
         bool "Ethernet Bonding driver Vendor Hooks"
         depends on VENDOR_HOOKS && BONDING
diff --git a/drivers/hooks/vendor_hooks.c b/drivers/hooks/vendor_hooks.c
index 7cd27de79a6b..d4251038740d 100644
--- a/drivers/hooks/vendor_hooks.c
+++ b/drivers/hooks/vendor_hooks.c
@@ -21,38 +21,38 @@
  */
 
 #ifdef CONFIG_VENDOR_BOND_HOOKS
-EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_bond_check_dev_link);
+EXPORT_HOOK_SYMBOL_GPL(vendor_bond_check_dev_link);
 #endif
 #ifdef CONFIG_VENDOR_LIBATA_HOOKS
 /* power on/off sata */
-EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_sata_reset);
+EXPORT_HOOK_SYMBOL_GPL(vendor_sata_reset);
 #endif
 
 #ifdef CONFIG_VENDOR_SPINOR_HOOKS
 /* change erase size by wanted_size */
-EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_spinor_erase_size);
+EXPORT_HOOK_SYMBOL_GPL(vendor_spinor_erase_size);
 /* return default info table to do sfdp */
-EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_spi_nor_read_id);
+EXPORT_HOOK_SYMBOL_GPL(vendor_spi_nor_read_id);
 #endif
 
 #ifdef CONFIG_VENDOR_IGB_HOOKS
-EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_igb_init);
-EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_igb_random_mac);
-EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_igb_eth_name);
-EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_igb_remove);
-EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_igb_open);
-EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_igb_close);
-EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_igb_get_skb);
-EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_igb_ioctl);
-EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_igb_set_spd_dplx);
-EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_igb_phy_force_speed_duplex_setup);
+EXPORT_HOOK_SYMBOL_GPL(vendor_igb_init);
+EXPORT_HOOK_SYMBOL_GPL(vendor_igb_random_mac);
+EXPORT_HOOK_SYMBOL_GPL(vendor_igb_eth_name);
+EXPORT_HOOK_SYMBOL_GPL(vendor_igb_remove);
+EXPORT_HOOK_SYMBOL_GPL(vendor_igb_open);
+EXPORT_HOOK_SYMBOL_GPL(vendor_igb_close);
+EXPORT_HOOK_SYMBOL_GPL(vendor_igb_get_skb);
+EXPORT_HOOK_SYMBOL_GPL(vendor_igb_ioctl);
+EXPORT_HOOK_SYMBOL_GPL(vendor_igb_set_spd_dplx);
+EXPORT_HOOK_SYMBOL_GPL(vendor_igb_phy_force_speed_duplex_setup);
 #endif
 
 #ifdef CONFIG_VENDOR_I40E_HOOKS
-EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_i40e_rebuild);
-EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_i40e_netdev_stats);
+EXPORT_HOOK_SYMBOL_GPL(vendor_i40e_rebuild);
+EXPORT_HOOK_SYMBOL_GPL(vendor_i40e_netdev_stats);
 #endif
 
 #ifdef CONFIG_VENDOR_INET_HOOKS
-EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_ip_finish_output);
+EXPORT_HOOK_SYMBOL_GPL(vendor_ip_finish_output);
 #endif
diff --git a/include/trace/hooks/vendor_hooks.h b/include/trace/hooks/vendor_hooks.h
index ab8864da66d8..8264d33698c0 100644
--- a/include/trace/hooks/vendor_hooks.h
+++ b/include/trace/hooks/vendor_hooks.h
@@ -9,10 +9,15 @@
 
 #include <linux/tracepoint.h>
 
-#if defined(CONFIG_TRACEPOINTS) && defined(CONFIG_VENDOR_HOOKS)
+#ifdef CONFIG_VENDOR_HOOKS
+
+#ifdef CONFIG_TRACEPOINTS
 
 #define DECLARE_HOOK DECLARE_TRACE
 
+#define EXPORT_HOOK_SYMBOL_GPL(name)					\
+	EXPORT_TRACEPOINT_SYMBOL_GPL(name)
+
 #ifdef TRACE_HEADER_MULTI_READ
 
 #define DEFINE_HOOK_FN(_name, _reg, _unreg, proto, args)		\
@@ -111,9 +116,81 @@
 
 #endif /* TRACE_HEADER_MULTI_READ */
 
-#else /* !CONFIG_TRACEPOINTS || !CONFIG_VENDOR_HOOKS */
+#else /* !CONFIG_TRACEPOINTS */
+
+#ifndef _VENDOR_HOOKS_H
+#define _VENDOR_HOOKS_H
+
+struct vendor_hook {
+	struct static_key key;
+	void __rcu *func;
+	void *data;
+};
+
+#define EXPORT_HOOK_SYMBOL_GPL(name)					\
+	struct vendor_hook __vendorhook_##name = {			\
+		.key = STATIC_KEY_INIT_FALSE,				\
+		.func = NULL,						\
+		.data = NULL };						\
+	EXPORT_SYMBOL_GPL(__vendorhook_##name)
+
+#define __DO_VENDOR_HOOK(name, args, data_proto)			\
+	do {								\
+		void (*probe)(data_proto) =				\
+			rcu_dereference_raw(__vendorhook_##name.func);	\
+		probe(__vendorhook_##name.data, TP_ARGS(args));		\
+	} while (0)
+
+#define __DECLARE_HOOK(name, proto, args, data_proto)			\
+	extern struct vendor_hook __vendorhook_##name;			\
+	static inline void trace_##name(proto)				\
+	{								\
+		if (static_key_false(&__vendorhook_##name.key))		\
+			__DO_VENDOR_HOOK(name, TP_ARGS(args),		\
+					 PARAMS(data_proto));		\
+	}								\
+	static inline int						\
+	register_trace_##name(void (*probe)(data_proto), void *data)	\
+	{								\
+		rcu_assign_pointer(__vendorhook_##name.func, probe);	\
+		__vendorhook_##name.data = data;			\
+		static_key_enable(&__vendorhook_##name.key);		\
+		return 0;						\
+	}								\
+	static inline int						\
+	unregister_trace_##name(void (*probe)(data_proto), void *data)	\
+	{								\
+		static_key_disable(&__vendorhook_##name.key);		\
+		rcu_assign_pointer(__vendorhook_##name.func, NULL);	\
+		__vendorhook_##name.data = NULL;			\
+		return 0;						\
+	}								\
+	static inline bool						\
+	trace_##name##_enabled(void)					\
+	{								\
+		return static_key_false(&__vendorhook_##name.key);	\
+	}
+
+#ifdef CREATE_TRACE_POINTS
+#undef CREATE_TRACE_POINTS
+
+#undef DECLARE_HOOK
+#define DECLARE_HOOK(name, proto, args)
+
+#else /* !CREATE_TRACE_POINTS */
+
+#define DECLARE_HOOK(name, proto, args)					\
+	__DECLARE_HOOK(name, PARAMS(proto), PARAMS(args),		\
+		       PARAMS(void *__data, proto))
+#endif
+
+#endif /* _VENDOR_HOOKS_H */
+
+#endif /* CONFIG_TRACEPOINTS */
+
+#else /* !CONFIG_VENDOR_HOOKS */
 /* suppress trace hooks */
 #define DECLARE_HOOK DECLARE_EVENT_NOP
 #define DECLARE_RESTRICTED_HOOK(name, proto, args, cond)		\
 	DECLARE_EVENT_NOP(name, PARAMS(proto), PARAMS(args))
-#endif
+#endif /* CONFIG_VENDOR_HOOKS */
-- 
2.17.1

